### PR TITLE
gvk partial inference

### PR DESCRIPTION
**Pull request by wangxinyi7**
### Description

In order to interact with the resource API, we need to type the full group.version.kind, `consul resource read catalog.v2beta1.Node`, which is cumbersome.

In this PR, the resource CLI is able to infer the full resource GVK format from the resource kind. Here is a highlight list of the CLI behaviour:

- **it uses local copy of registrations to build the inference map**
- the kind is case insensitive
- if there are several resources that share the same `kind` name, it throws error to enforce the user to use the full gvk format. You could try `demo.v1.Artist` and `demo.v2.Artist`.

**Examples:**
- `consul resource read node` is equivalent to `consul resource read catalog.v2beta1.Node`
- You can not use `consul resource read album` since there are two kinds of album: `demo.v1.Album` and `demo.v2.Album`

Since we are using local copy, we only support this feature in read&list commands and didn't touch delete&apply commands for safety reason. After we decouple the CLI command from the consul server, we will fully support this feature.

### PR Checklist

* ~[ ] updated test coverage~
* [ ] external facing docs updated
* [ ] appropriate backport labels added
* ~[ ] not a security concern~

### CODE REVIEW

**Comment by david-yu** [comment]
@wangxinyi7 Could you provide examples of before and after of how this would work with a Consul resource that is currently available?

**Comment by david-yu** [comment]
Thanks for the clarification @wangxinyi7. Would it be also possible to include docs changes here in https://github.com/hashicorp/consul/pull/19050 to describe new CLI behavior with resources and how we can infer gvk references perhaps after this is done? I would like to see concrete examples listed in our docs on how to list these Consul resources we are going to ship as part of 1.17. cc @boruszak

**Comment by wangxinyi7** [comment]
> Thanks for the clarification @wangxinyi7. Would it be also possible to include docs changes here in #19050 to describe new CLI behavior with resources and how we can infer gvk references perhaps after this is done? I would like to see concrete examples listed in our docs on how to list these Consul resources we are going to ship as part of 1.17. cc @boruszak

@david-yu yeah we for sure need to update docs, sounds good, will update docs after this is done 👍

File: command/resource/helper.go [hunk 1]
```diff
@@ -233,3 +225,51 @@ func (resource *Resource) List(gvk *GVK, q *client.QueryOptions) (*ListResponse,
...
+	if len(s) != 3 {
+		return nil, fmt.Errorf("Must provide resource type argument with either in group.verion.kind format or its shorthand name")
+	}
+
+	return &GVK{
+		Group:   s[0],
+		Version: s[1],
+		Kind:    s[2],
+	}, nil
+}
```
Comment (dhiaayachi) [inline command/resource/helper.go:258]: I think this is a great candidate to be converted to a switch statement

**Review by dhiaayachi — COMMENTED** [review]
Sorry for the late review @wangxinyi7 😞 !
I left few comments but I have some reservation around the global approach here:
The command will execute locally in the CLI binary, which could lead to a discrepancy between what the consul api-resources return and what the server actually support. We also do the inference locally in the CLI which also could lead to the same problem.

To illustrate the problem, let's say we have a `demo.v1.foo` type in 1.17.x and we add `demo.v2.foo` in 1.18.x. Now if you execute a command using 1.17.x cli binary and giving `foo` against an 1.18.x server `demo.v1.foo` will be inferred while the correct behaviour should be to return an error as we have 2 foo types.

**Comment by wangxinyi7** [comment]
> Sorry for the late review @wangxinyi7 😞 ! I left few comments but I have some reservation around the global approach here: The command will execute locally in the CLI binary, which could lead to a discrepancy between what the consul api-resources return and what the server actually support. We also do the inference locally in the CLI which also could lead to the same problem.
> 
> To illustrate the problem, let's say we have a `demo.v1.foo` type in 1.17.x and we add `demo.v2.foo` in 1.18.x. Now if you execute a command using 1.17.x cli binary and giving `foo` against an 1.18.x server `demo.v1.foo` will be inferred while the correct behaviour should be to return an error as we have 2 foo types.

Hi @dhiaayachi That's a reaaaaally great catch! Thanks! Let me think how to resolve this issue.

**Comment by dhiaayachi** [comment]
@wangxinyi7 I think the only option here would be to perform the inference at the server side more pricesily in the `resourceServiceServer` level. So we can be sure that the inferred gvk is correct. 
I think generating the api-resources locally could be fine as long as it's documented, at least for now, but  I think we should run this by a larger audience.

File: command/resource/helper.go [hunk 2]
```diff
@@ -120,19 +121,10 @@ func GetTypeAndResourceName(args []string) (gvk *GVK, resourceName string, e err
 	if strings.HasPrefix(args[1], "-") {
 		return nil, "", fmt.Errorf("Must provide resource name right after type")
 	}
+	resourceName = args[1]
```
[the remark this answers is not in the document]
Comment (wangxinyi7) [reply command/resource/helper.go:124]: This one is already sanitized before calling this function:
[read](https://github.com/hashicorp/consul/blob/e6ed18155154e8674881d0a6a867977f3649f434/command/resource/read/read.go#L89)
[delete](https://github.com/hashicorp/consul/blob/e6ed18155154e8674881d0a6a867977f3649f434/command/resource/delete/delete.go#L87)

File: command/resource/helper.go (hunk 1 above)
Comment (wangxinyi7) [reply command/resource/helper.go:258]: Right! used switch to make the statements more readable

File: command/resource/helper.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhiaayachi) [reply command/resource/helper.go:158]: In that case I think it's better if we change the signature of this func to not take a slice but 2 arguments.

File: command/resource/helper.go [hunk 3]
```diff
@@ -282,3 +259,54 @@ func (resource *Resource) List(gvk *GVK, q *client.QueryOptions) (*ListResponse,
...
+		}
+	case length != 3:
+		return nil, fmt.Errorf("Must provide resource type argument with either in group.verion.kind format or its shorthand name")
+	}
+
+	return &GVK{
+		Group:   s[0],
+		Version: s[1],
+		Kind:    s[2],
+	}, nil
```
Comment (dhiaayachi) [inline command/resource/helper.go:295]:
```suggestion
	switch len(s); {
	// only kind is provided
	case 1:
		kindToGVKMap := BuildKindToGVKMap()
		kind := strings.ToLower(s[0])
		switch len(kindToGVKMap[kind]) {
		// no g.v.k is found
		case 0:
			return nil, fmt.Errorf("The shorthand name does not map to any existing resource type, please check `consul api-resources`")
		// only one is found
		case 1:
			// infer gvk from resource kind
			gvkSplit := strings.Split(kindToGVKMap[kind][0], ".")
			return &GVK{
				Group:   gvkSplit[0],
				Version: gvkSplit[1],
				Kind:    gvkSplit[2],
			}, nil
		// it alerts error if any conflict is found
		default:
			return nil, fmt.Errorf("The shorthand name has conflicts %v, please use the full name", kindToGVKMap[s[0]])
		}
        case 3:
                return &GVK{
		   Group:   s[0],
		   Version: s[1],
		   Kind:    s[2],
	        }, nil
	default:
		return nil, fmt.Errorf("Must provide resource type argument with either in group.verion.kind format or its shorthand name")
	}
```

**Review by dhiaayachi — APPROVED** [review]
Hey @wangxinyi7 
I added few minor comments that would simplify the code but overall LGTM!

File: command/resource/helper.go (hunk 3 above)
Comment (wangxinyi7) [reply command/resource/helper.go]: Thanks, this one looks much better!

File: command/resource/helper.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wangxinyi7) [reply command/resource/helper.go:161]: I tried to re-crafted the function signature to receive `GetTypeAndResourceName(resourceType string, resourceName string)` then I found I need to do the sanity check 

> if strings.HasPrefix(args[1], "-") {

at the caller side which shifted more code the caller side. So I moved the arguments sanity check inside this function.

**Comment by david-yu** [comment]
@wangxinyi7 Should this have been backported to 1.17.x?

**Pull request merged**